File: 345-reverse-vowels-of-a-string/345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    bool inside(char c){
        if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
           c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U')
            return true;
        return false;
    }
    
    string reverseVowels(string s) {
        int left = 0, right = s.size() - 1;
        bool l_found = false, r_found = false;
        while (left < right)
        {
            if (!l_found)
            {
                if (!inside(s[left])){
                    left++;
                }
                else
                    l_found = true;
            }
            
            if (!r_found)
            {
                if (!inside(s[right])){
                    right--;
                }
                else
                    r_found = true;
            }
            
            if (l_found && r_found){
                swap(s[left], s[right]);
                l_found = false;
                r_found = false;
                left++; right--;
            }
            // cout<<left<<" "<<right<<" "<<s<<endl;
        }
        return s;
    }
};

File: 345-reverse-vowels-of-a-string/345-reverse-vowels-of-a-string.rb
# @param {String} s
# @return {String}
def reverse_vowels(s)
    vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
    left, right = 0, s.length - 1
    while left < right
       while left < right and not vowels.include?(s[left]) 
          left += 1
       end
        
       while left < right and not vowels.include?(s[right]) 
          right -= 1
       end
        
       s[left], s[right] = s[right], s[left]
       left += 1
       right -= 1
    end
    s
end